File: src/main/java/gargoyle/xenox/util/res/audio/AudioClipImpl.java
package gargoyle.xenox.util.res.audio;

import javax.sound.sampled.AudioInputStream;
import javax.sound.sampled.AudioSystem;
import javax.sound.sampled.Clip;
import javax.sound.sampled.LineUnavailableException;
import javax.sound.sampled.UnsupportedAudioFileException;
import java.io.IOException;
import java.io.InputStream;
import java.net.URL;

class AudioClipImpl implements AudioClip {
    private final Clip clip;

    public AudioClipImpl(URL location) {
        Clip clip = null;
        try {
            try (InputStream stream = location.openConnection().getInputStream();
                 AudioInputStream sound = AudioSystem.getAudioInputStream(stream)) {
                clip = AudioSystem.getClip();
                clip.open(sound);
            }
        } catch (IOException | UnsupportedAudioFileException | LineUnavailableException e) {
            clip = null;
            throw new RuntimeException(e.getLocalizedMessage(), e);
        } finally {
            this.clip = clip;
        }
    }

    @Override
    public void play() {
        if (clip != null && clip.isOpen()) {
            clip.setFramePosition(0);
            clip.start();
        }
    }

    @Override
    public void loop() {
        if (clip != null && clip.isOpen()) {
            clip.loop(Clip.LOOP_CONTINUOUSLY);
        }
    }

    @Override
    public void stop() {
        if (clip != null && clip.isOpen()) {
            clip.stop();
        }
    }

    @Override
    public void close() {
        if (clip != null && clip.isOpen()) {
            clip.close();
        }
    }
}

File: src/main/kotlin/gargoyle/xenox/util/res/audio/AudioClipImpl.kt
package gargoyle.xenox.util.res.audio

import java.io.BufferedInputStream
import java.net.URL
import javax.sound.sampled.AudioSystem
import javax.sound.sampled.Clip

internal class AudioClipImpl(location: URL) : AudioClip {
    private val clip: Clip

    init {
        clip = AudioSystem.getClip().also {
            location.openConnection().getInputStream().use { stream ->
                BufferedInputStream(stream).use { bufferedInputStream ->
                    AudioSystem.getAudioInputStream(bufferedInputStream).use { sound ->
                        it.open(sound)
                    }
                }
            }
        }
    }

    override fun play() {
        if (clip.isOpen) {
            clip.framePosition = 0
            clip.start()
        }
    }

    override fun loop() {
        if (clip.isOpen) {
            clip.loop(Clip.LOOP_CONTINUOUSLY)
        }
    }

    override fun stop() {
        if (clip.isOpen) {
            clip.stop()
        }
    }

    override fun close() {
        if (clip.isOpen) {
            clip.close()
        }
    }
}
